src/librustc_mir/monomorphize/polymorphize.rs

```diff
@@ -0,0 +1,283 @@
+
+use rustc::mir::{
+    visit::{TyContext, Visitor},
+    CastKind, Local, LocalDecl, Location, ReadOnlyBodyAndCache, Rvalue, TerminatorKind,
+};
+use rustc::ty::{
+    self,
+    adjustment::PointerCast,
+    fold::{TypeFoldable, TypeVisitor},
+    query::Providers,
+    Ty, TyCtxt,
+};
+use rustc_hir::{self as hir, def_id::DefId};
+use rustc_index::bit_set::BitSet;
+use rustc_span::symbol::Symbol;
+use rustc_target::spec::abi::Abi;
+
+/// Provide implementations of queries relating to polymorphization analysis.
+pub fn provide(providers: &mut Providers<'_>) {
+    providers.used_generic_parameters = used_generic_parameters;
+}
+
+/// Determine which generic parameters are used by the function/method/closure represented by
+/// `def_id`. Returns a `BitSet` with a domain size matching the number of generic parameters,
+/// set bits represent that the matching generic parameter is used.
+fn used_generic_parameters(tcx: TyCtxt<'_>, def_id: DefId) -> BitSet<u32> {
+    debug!("used_generic_parameters({:?})", def_id);
+    let generics = tcx.generics_of(def_id);
+    debug!("used_generic_parameters: generics={:?}", generics);
+
+    if !tcx.sess.opts.debugging_opts.polymorphize {
+        // If the analysis is not enabled, then return a `BitSet` saying that every generic
+        // parameter was used.
+        return BitSet::new_filled(generics.count());
+    }
+
+    // If the item is not local then say every generic parameter was used.
+    if !def_id.is_local() {
+        return BitSet::new_filled(generics.count());
+    }
+
+    let hir_id = tcx.hir().as_local_hir_id(def_id).unwrap();
+    debug!("used_generic_parameters: hir_id={:?}", hir_id);
+
+    // Prohibit any node that would not result in a `MonoItem::Fn` and any non-Rust
+    // ABIs (incl. intrinsics).
+    let node = tcx.hir().get(hir_id);
+    debug!("used_generic_parameters: node={:?}", node);
+    match node {
+        hir::Node::Expr(&hir::Expr { kind: hir::ExprKind::Closure(..), .. }) => (),
+        hir::Node::Item(&hir::Item { kind: hir::ItemKind::Fn(..), .. })
+        | hir::Node::ImplItem(&hir::ImplItem { kind: hir::ImplItemKind::Method(..), .. })
+        | hir::Node::TraitItem(&hir::TraitItem {
+            kind: hir::TraitItemKind::Method(_, hir::TraitMethod::Provided(..)),
+            ..
+        }) => match tcx.fn_sig(def_id).abi() {
+            Abi::Rust => (),
+            _ => return BitSet::new_filled(generics.count()),
+        },
+        _ => return BitSet::new_filled(generics.count()),
+    }
+
+    let mut used_parameters = BitSet::new_empty(generics.count());
+    debug!(
+        "used_generic_parameters: used_parameters.domain_size()={:?}",
+        used_parameters.domain_size()
+    );
+    if generics.count() == 0 {
+        // Exit early when there are no parameters to be unused.
+        return used_parameters;
+    }
+
+    let leaf_is_closure =
+        if let hir::Node::Expr(&hir::Expr { kind: hir::ExprKind::Closure(..), .. }) = node {
```

Comment:
I think you can check using the `DefId` instead (likely a `tcx` method).

Reply:
I've fixed this by removing this line in fe5ba4d.